src/library/autodj/autodjprocessor.cpp

```diff
@@ -817,61 +972,204 @@ void AutoDJProcessor::calculateTransition(DeckAttributes* pFromDeck,
+            if (outroLength > introLength) {
+                pFromDeck->fadeBeginPos = outroEnd - introLength;
+                pFromDeck->fadeEndPos = outroEnd;
+                pToDeck->startPos = introStart;
+            } else {
+                pFromDeck->fadeBeginPos = outroStart;
+                pFromDeck->fadeEndPos = outroEnd;
+                pToDeck->startPos = introStart;
+            }
+        } else if (outroLength > 0 && introLength <= 0) {
+            pFromDeck->fadeBeginPos = outroStart;
+            pFromDeck->fadeEndPos = outroEnd;
+            pToDeck->startPos = introStart;
+        } else if (introLength > 0 && outroLength <= 0) {
+            pFromDeck->fadeBeginPos = outroEnd - introLength;
+            pFromDeck->fadeEndPos = outroEnd;
+            pToDeck->startPos = introStart;
         } else {
-            pFromDeck->posThreshold = 1.0;
-            pFromDeck->fadeDuration = 0.0;
+            useFixedFadeTime(pFromDeck, pToDeck,
+                    getLastSoundPosition(pFromDeck),
+                    getFirstSoundPosition(pToDeck));
         }
-
-        // We don't set pToDeck->startPos here because it is only needed
-        // for pause transition.
-    } else {  // Pause transition (transition time <= 0)
-        if (fromTrackDuration > 0.0) {
-            pFromDeck->posThreshold = fromTrackOutroEnd / fromTrackDuration;
-            pFromDeck->fadeDuration = m_transitionTime / fromTrackDuration;
+        break;
+    case TransitionMode::AlignIntroOutroStart:
+        // Use the outro or intro length for the transition time, whichever is
+        // shorter. If the outro is longer than the intro, cut off the end
+        // of the outro.
+        //
+        // In the diagrams below,
+        // - is part of a track outside the outro/intro,
+        // o is part of the outro
+        // i is part of the intro
+        // | marks the boundaries of the transition
+        //
+        // When outro > intro:
+        // ------|ooo|ooo
+        //       |iii|------
+        //
+        // When outro < intro:
+        // ------|ooo|
+        //       |iii|iii-----
+        //
+        // If only the outro or intro length is marked but not both, use the one
+        // that is marked for the transition time. If neither is marked, fall
+        // back to the transition time from the spinbox.
+        if (outroLength > 0 && introLength > 0) {
+            pFromDeck->fadeBeginPos = outroStart;
+            if (outroLength > introLength) {
+                // Cut off end of outro
+                pFromDeck->fadeEndPos = outroStart + introLength;
+            } else {
+                pFromDeck->fadeEndPos = outroEnd;
+            }
+            pToDeck->startPos = introStart;
+        } else if (outroLength > 0 && introLength <= 0) {
+            pFromDeck->fadeBeginPos = outroStart;
+            pFromDeck->fadeEndPos = outroEnd;
+            pToDeck->startPos = introStart;
```

Comment:
Here we must not seek, because introStart is not relevant if no intro end is set. 
In this case the track should not be seeked.

Reply:
introStart is still relevant to cut off silence

Reply:
If you wish to use this point you can select this as load point in the deck preferences.

In my case I have for many tracks a cue point set, which I like to use until I have defined an intro. So I would select load at cue point.

Reply:
Again we come to this migration issue. I do not want AutoDJ to have anything to do with the deck loading preference nor the main cue point. Doing that would make it less reliable and more confusing. The only solution I can think of that would not compromise the reliability of AutoDJ would be a button to set intro start points at the main cue point for every track in the library. If this would not give good results, then I do not understand why you would want AutoDJ to use the main cue point anyway.

Reply:
I can understand that you mode is more reliable finally, in case the user has re-prepared the tracks. Since this will unlikely happen at once we can tell with my proposed behavior a smooth migration story. Since we had not much issues with the old auto-DJ, most user migrating from 2.2 will expect that it continues to work as usual.

The 2.2 mode was start at the load point 0:00 or cue (deck preferences), intro and autro length = transition time and outo end at the very end. 

After migration and without additional actions, the outro end moves to the silence start point, which is a welcome change IMHO. But everything else the user is used to, will work and the next gig can be done without additional hassle.  

If the user likes to rely on the automatically intro in point (your case) he can switch the deck load options to that new option. If he likes to step by step prepare intros, he can do it and they are used instead of the load preferences. 

I am not sure if I will actually prepare many intros, because once I have played a track during prepare or live, it has automatically a cue point. So auto DJ is able to start the track where I have started it before.

If I do not like the behavior I can switch to intro start load point. 

I hope you see, how this makes sense and does not limit anyone.

Reply:
I think making the deck load preferences have anything to do with AutoDJ will confuse people.

> I am not sure if I will actually prepare many intros, because once I have played a track during prepare or live, it has automatically a cue point. So auto DJ is able to start the track where I have started it before.

How about adding an option to the Deck preferences to automatically move the intro start with the main cue point? That would make the intro start feature more accessible to users whose controllers do not have the intro & outro buttons mapped.

Reply:
> Again, if we do that, AutoDJ will not respect the intro start cues as it says it will because upgrading users would somehow have to figure out that they need to change a new preference option.

The preferences option is already there in Mixxx 2.3 and it was always respected by the Auto Dj. Not respecting it would be a probably supprising for old users. For me an undesired regression from a working start at cue mode.

I think a way to keep this behaviour by default is somehow mandatory for me.

> How about adding an option to the Deck preferences to automatically move the intro start with the main cue point?

On first sight a good idea. But I find it also interesting to define a intro independent from the cue points.
How can I detach the intro start point in this case? When intro start end end is defined? That would work for me.

Using the cue button to define intro start sounds great. I will probably (continue to) use it.

If we manage to make the migration that all "load at cue setups" have this feature enabled by default it will work for me. The "load at start" guys can use load at intro start, because it is load at sound start by default.
They did not care about cue as auto DJ start before so they will not do now.

Reply:
> On first sight a good idea. But I find it also interesting to define a intro independent from the cue points.
How can I detach the intro start point in this case? When intro start end end is defined?

IMO that would be too complicated and I anticipate many users would get confused. Either use the cue point to move intro start or do not. 

> If we manage to make the migration that all "load at cue setups" have this feature enabled by default it will work for me.

I think this option should be off by default. I do not think we should assume that because someone has that option enabled they intend to use the main cue as the AutoDJ start point. Considering the option for deck loading at the main cue has been on by default, I do not think we should assume that users are even aware the option exists if they have it at the default setting. For users who have set their main cue points with the intention of having AutoDJ start playing from there, it will be good to have a smooth migration path. However, I suspect most users do not do this, so I think it is fine to tell users that have been using the main cue as the AutoDJ start point to do a few more clicks in the preferences.